src/sandstorm/indexer.capnp

```diff
@@ -0,0 +1,168 @@
+  # Capability used to index the content of a grain.
+  #
+  # This is a one-way capability. Although GrainIndexer is implemented by the indexer and is called
+  # by arbitrary content grains, Sandstorm ensures that information cannot leak from the indexer to
+  # the content grains by implementing a one-way message queue in between. This ensures that bugs
+  # (or malicious backdoors) in the indexer cannot be exploited by an app to learn about other
+  # information in the index.
+  #
+  # One-way communication is enforced via a message queue: When a content grain calls index(),
+  # Sandstorm places the call parameters in a queue and returns immediately. Later on, Sandstorm
+  # takes calls from the queue and actually delivers them to the indexer. This means that not only
+  # is the indexer unable to return data to the caller, but the caller cannot find out how long it
+  # takes to process the call nor if it threw an exception.
+  #
+  # If data passed to `index()` contains capabilities, Sandstorm seals those capabilities such that
+  # the indexer cannot communicate to the outside world through them. In the case of `UiView`
+  # capabilities, the indexer receives a sealed `UiView` which can only be used in limited ways:
+  # * It can be `save()`ed.
+  # * It can be passed to `offer()`, causing the grain to open in the user's UI. However, Sandstorm
+  #   implements this such that revoking the user's access to the indexer grain does not cause them
+  #   to lose access to capabilities obtained through it, since that wouldn't make sense for this
+  #   use case.
+  # * Its `getViewInfo()` method can be called. However, the results will be served from cache
+  #   without invoking the underlying grain.
+
+  index @0 (path :Text, content :IndexableContent(Metadata));
+  # Add content of the given path to the index.
+  #
+  # A null value for `content` indicates that the path doesn't exist.
+  #
+  # TODO(someday): When Cap'n Proto supports bulk-transfer methods with flow control, mark this
+  #   method as such. This will cause the Cap'n Proto implementation to pretend the method has
+  #   completed (resolving the promise) as soon as it's appropriate for the caller to make another
+  #   call. For now, apps should pretend this is already in place, and make only one call at a
+  #   time -- the current performance penalty in doing so probably isn't a big deal.
+}
+
+struct IndexableContent(Metadata) {
+  body @0 :Text;
+  # Freeform natural-language body text. Will be tokenized by the search indexer.
+
+  links @1 :List(Grain.UiView);
+  # Other grains that can be accessed through this content.
+
+  threadPath @2 :Text;
+  # Optional path of thread of which this item is a part, if any. See `ActivityEvent.thread`.
+
+  metadata @3 :Metadata;
+  # Metadata for search operators, e.g. "subject:", "author:", etc. Each app can define its own
+  # metadata format.
+  #
+  # TODO(someday): Spec out how this works. Not needed for MVP of search.
+}
```

Comment:
It's not obvious to me how an indexer would be able to use "generic" operators like this while treating the metadata as completely opaque.

If the plan is to not implement this right away anyway, is there a good reason to even include this method? We can always add more later...